commit 5c46d9a4c85abbcc0b5db2bbbafded3efd784cd9
Author: Jiang Jiang <jiang.jiang@nokia.com>
Date:   Thu Nov 25 17:30:32 2010 +0100

    Take Xft.hintstyle by default to match the behavior of GTK+
    
    For Qt apps running in GNOME, we use the GTK+ settings by
    default (instead of fontconfig settings).
    
    Task-number: QTBUG-13800
    Reviewed-by: Samuel

---
 src/gui/kernel/qapplication_x11.cpp |    4 ++++
 src/gui/text/qfontengine_x11.cpp    |    7 ++++++-
 2 files changed, 10 insertions(+), 1 deletion(-)

--- a/src/gui/kernel/qapplication_x11.cpp
+++ b/src/gui/kernel/qapplication_x11.cpp
@@ -1589,6 +1589,9 @@ static void getXDefault(const char *grou
         int v = strtol(str, &end, 0);
         if (str != end)
             *val = v;
+        // otherwise use fontconfig to convert the string to integer
+        else
+            FcNameConstant((FcChar8 *) str, val);
     }
 }
 
@@ -2233,6 +2236,7 @@ void qt_init(QApplicationPrivate *priv,
         }
         getXDefault("Xft", FC_ANTIALIAS, &X11->fc_antialias);
 #ifdef FC_HINT_STYLE
+        X11->fc_hint_style = -1;
         getXDefault("Xft", FC_HINT_STYLE, &X11->fc_hint_style);
 #endif
 #if 0
--- a/src/gui/text/qfontengine_x11.cpp
+++ b/src/gui/text/qfontengine_x11.cpp
@@ -1015,7 +1015,12 @@ QFontEngineX11FT::QFontEngineX11FT(FcPat
 #ifdef FC_HINT_STYLE
     {
         int hint_style = 0;
-        if (FcPatternGetInteger (pattern, FC_HINT_STYLE, 0, &hint_style) == FcResultNoMatch)
+        // Try to use Xft.hintstyle from XDefaults first if running in GNOME, to match
+        // the behavior of cairo
+        if (X11->fc_hint_style > -1 && X11->desktopEnvironment == DE_GNOME)
+            hint_style = X11->fc_hint_style;
+        else if (FcPatternGetInteger (pattern, FC_HINT_STYLE, 0, &hint_style) == FcResultNoMatch
+                 && X11->fc_hint_style > -1)
             hint_style = X11->fc_hint_style;
 
         switch (hint_style) {
